consensus/src/pipeline/header_processor/pre_pow_validation.rs

```diff
@@ -0,0 +1,198 @@
+        let mut missing_parents = Vec::new();
+        for parent in header.direct_parents() {
+            match self
+                .statuses_store
+                .read()
+                .get(*parent)
+                .unwrap_option()
+            {
+                None => missing_parents.push(*parent),
+                Some(StatusInvalid) => {
+                    return Err(RuleError::InvalidParent(*parent));
+                }
+                Some(_) => {}
+            }
+        }
+        if !missing_parents.is_empty() {
+            return Err(RuleError::MissingParents(missing_parents));
+        }
+        Ok(())
+    }
+
+    fn check_parents_incest(
+        self: &Arc<HeaderProcessor>, ctx: &mut HeaderProcessingContext, header: &Header,
+    ) -> BlockProcessResult<()> {
+        let parents = ctx.non_pruned_parents();
+        for parent_a in parents.iter() {
+            for parent_b in parents.iter() {
+                if parent_a == parent_b {
+                    continue;
+                }
+
+                if self
+                    .reachability_service
+                    .is_dag_ancestor_of(*parent_a, *parent_b)
+                {
+                    return Err(RuleError::InvalidParentsRelation(*parent_a, *parent_b));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn check_pruning_violation(
+        self: &Arc<HeaderProcessor>, ctx: &mut HeaderProcessingContext, header: &Header,
+    ) -> BlockProcessResult<()> {
+        match self
+            .pruning_store
+            .read()
+            .pruning_point()
+            .unwrap_option()
+        {
+            None => Ok(()), // It implictly means that genesis is the pruning point - so no violation can exist
+            Some(pruning_point) => {
+                let non_pruned_parents = ctx.non_pruned_parents();
+                if non_pruned_parents.is_empty() {
+                    return Ok(());
+                }
+
+                if non_pruned_parents.iter().cloned().any(|parent| {
+                    !self
+                        .reachability_service
+                        .is_dag_ancestor_of(pruning_point, parent)
+                }) {
+                    return Err(RuleError::PruningViolation(pruning_point));
+                }
+
+                Ok(())
+            }
+        }
+    }
+
+    fn check_pow(
+        self: &Arc<HeaderProcessor>, ctx: &mut HeaderProcessingContext, header: &Header,
+    ) -> BlockProcessResult<()> {
+        Ok(()) // TODO: Check PoW
+    }
+
+    fn check_difficulty_and_daa_score(
+        self: &Arc<HeaderProcessor>, ctx: &mut HeaderProcessingContext, header: &Header,
+    ) -> BlockProcessResult<()> {
+        let ghostdag_data = ctx.ghostdag_data.clone().unwrap();
+        let window = self
+            .dagtraversal_manager
+            .block_window(ghostdag_data, self.difficulty_window_size);
+
+        ctx.block_window = Some(window.clone());
```

Comment:
I think you should use the window throughout this function and *move* it to the ctx without cloning at the end of the function